fix(chatform): Broaden URL matching to any string following scheme

anthonybilinski opened the pull request:
Fix #4853

Instead of searching strictly valid URLs, allow anything characters following scheme. This allows for UTF-8 characters used in other lanagues, as well as parentheses and other ASCII encoding. This will over-match some invalid URLs.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/4951)
<!-- Reviewable:end -->

tox-user commented:
A few more types of characters should be excluded from the url. 
Trailing `?!:`
and when the link is inside of ( ) or [ ] or quotes those surrounding characters should not be matched either. So I guess that also means trailing `) ] ' "`

tox-user commented:
Examples:
`(https://en.wikipedia.org/wiki/Tox_(protocol))` - trailing ) will be matched
`list of voip programs from https://en.wikipedia.org/wiki/Comparison_of_VoIP_software:` - trailing : will be matched
`href="https://tox.chat"` - the last " will be matched

anthonybilinski commented:
Good comments, have addressed them in the new version. Parsing ignores matching pairs of brackets etc. as well as ending punctuation, without ignoring brackets that are actually part of the URL. This doesn't add punycode on hover, although I think that would be a good feature.

anthonybilinski commented:
Just realized there's some overlap in functionality with the existing PR https://github.com/qTox/qTox/pull/4297 which has been open for months, it sounds like this addressed most of the comments from that as well.

sudden6 commented:
@tox-user could you add the examples you gave to the test suite? This way we can ensure it doesn't get broken on accident.

tox-user commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/chatlog/textformatter.cpp, line 81 at r1](https://reviewable.io:443/reviews/qtox/qtox/4951#-L5eZE2wMYTr0rCVRTvw:-L5eZE2wMYTr0rCVRTvx:bu8tdse) ([raw file](https://github.com/qtox/qtox/blob/b29377e8d103ad1dae3864e3f17a8123313eebf0/src/chatlog/textformatter.cpp#L81)):*
>  QRegularExpression(QStringLiteral(R"((?<=^|[\s\n])[\S]*((www)\S+))"))

What does this part of regex do? `(?<=^|[\s\n])[\S]*`
`\s` should match all whitespace which also includes `\n`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4951)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
@tox-user Yes, it should, but AFAIK it doesn't

anthonybilinski commented:
Tested [\s\n] vs \s and they both act equivalently (as they should according to regex spec), so removed the \n and [], updated PR.
>What does this part of regex do? (?<=^|[\s\n])[\S]*

`?<=` is a "positive lookbehind" https://www.regular-expressions.info/lookaround.html
it requires that the preceding character in this case matches either `^` (start of line) or (now) `\s`, whitespace. Then `[\S]*` means 0 or more of any non-whitespace characters, used to match leading characters like `(["` before the start of the scheme of the URI.

anthonybilinski committed:
fix(chatform): Broaden URL matching to include unicode

Fix #4853
Fix #4295

*Instead of searching strictly valid URIs, allow anything characters following scheme. This allows for UTF-8 characters used in other languages, as well as parentheses and other ASCII characters. This will over-match some invalid URLs.
*Ignored surrounding characters of URIs and ending punctuation
*Fix www-only links by adding http scheme to href

anthonybilinski commented:
I added tests, rebased to master tip, and fixed a bug I found with the tests where ending punctuation wouldn't be stripped if there were no surrounding characters :)

Diadlo commented:
`tox:0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01234567890A` doesn't higlighted for me. Do you have any ideas?

anthonybilinski commented:
It was because we had `\\d` inside of a QStringLiteral for the ToxId matching. QStringLiteral already escapes the `\` so having a double `\\` breaks it. Fixed and updated PR, ToxId's didn't seem to match before this change either, but do now.

Diadlo commented:
Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/chatlog/textformatter.cpp, line 88 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BAuNZRXcm4CiSP8dq:-L6BAuNZRXcm4CiSP8dr:b9as5bp) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L88)):*
> ```cpp
>     // Note: This does not match only strictly valid URLs, but we broaden search to any string following scheme to
>     // allow UTF-8 "IRI"s instead of ASCII-only URLs
>     QRegularExpression(QStringLiteral(R"((?<=^|\s)\S*((((http[s]?)|ftp)://)\S+))")),
> ```

Why do you think about splitting http[s] and ftp on 2 regeps? Or merge with `file|smb`

---

*[src/chatlog/textformatter.cpp, line 89 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BACsDuaSowtjK8VJX:-L6BACsDuaSowtjK8VJY:b-dg27kw) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L89)):*
> ```cpp
>     // allow UTF-8 "IRI"s instead of ASCII-only URLs
>     QRegularExpression(QStringLiteral(R"((?<=^|\s)\S*((((http[s]?)|ftp)://)\S+))")),
>     QRegularExpression(QStringLiteral(R"((?<=^|\s)\S*((file|smb)://([\S| ]*)))")),
> ```

Seems `[\S| ]` is equivalent to `\S`. Isn't it?

---

*[src/chatlog/textformatter.cpp, line 132 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BA_4WLaySUHmidgvP:-L6BA_4WLaySUHmidgvQ:b-sk8x6q) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L132)):*
> ```cpp
>     do {
>         matchFound = false;
>         for (QPair<QString, QString> const& surroundChars : URI_WRAPPING_CHARS)
> ```

Probably `auto` will be enought

---

*[src/chatlog/textformatter.cpp, line 156 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BAhF3oVqEDdN5MpzZ:-L6BAhF3oVqEDdN5Mpz_:bjdxo3y) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L156)):*
> ```cpp
>     if (startOfBareUri != curValidationStartPos) {
>         strippedMatch.valid = false;
>     }
> ```

`} else {`

---

*[src/chatlog/textformatter.cpp, line 185 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BAjZNRVBNU2zGyg_E:-L6BAjZNRVBNU2zGyg_F:b-ukjbp7) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L185)):*
> ```cpp
>             MatchingUri matchUri = stripSurroundingChars(match.capturedRef(uriWithWrapMatch),
>                    match.capturedStart(uriWithoutWrapMatch) - match.capturedStart(uriWithWrapMatch));
>             if (matchUri.valid == false) {
> ```

`!matchUri.valid`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4951)*
<!-- Sent from Reviewable.io -->

anthonybilinski commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/chatlog/textformatter.cpp, line 81 at r1](https://reviewable.io:443/reviews/qtox/qtox/4951#-L5eZE2wMYTr0rCVRTvw:-L6BK8wWtEKI3_x1zflH:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/b29377e8d103ad1dae3864e3f17a8123313eebf0/src/chatlog/textformatter.cpp#L81)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

>  QRegularExpression(QStringLiteral(R"((?<=^|[\s\n])[\S]*((www)\S+))"))
What does this part of regex do? `(?<=^|[\s\n])[\S]*`
`\s` matches all whitespace which also includes `\n`.
</blockquote></details>

Done.

---

*[src/chatlog/textformatter.cpp, line 88 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BAuNZRXcm4CiSP8dq:-L6BKbABdLoGPaG989pI:bqwydxc) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L88)):*
> Or merge with file|smb
As posted below, http/https have stricter criteria (no whitespace) than file/smb below. We could split all 4 into their own rules, but to me it seems like more duplication. If the only difference between different rules are the name of the schemes, combining rules with identical formats seems nicer to me.

---

*[src/chatlog/textformatter.cpp, line 89 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BACsDuaSowtjK8VJX:-L6BKL2h4S7RJb7zF6zF:bth5p94) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L89)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Seems `[\S| ]` is equivalent to `\S`. Isn't it?
</blockquote></details>

this is \S (uppercase S) not \s, which is the opposite. [\S| ] is saying "any non-whitespace character or the space character" since apparently file and smb URIs can have spaces in them.

---

*[src/chatlog/textformatter.cpp, line 132 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BA_4WLaySUHmidgvP:-L6BKzgyENPr_WQXc6v7:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L132)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Probably `auto` will be enought
</blockquote></details>

Done.

---

*[src/chatlog/textformatter.cpp, line 156 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BAhF3oVqEDdN5MpzZ:-L6BLEE9GwhJAgm29LIf:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L156)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`} else {`
</blockquote></details>

Done.

---

*[src/chatlog/textformatter.cpp, line 185 at r2](https://reviewable.io:443/reviews/qtox/qtox/4951#-L6BAjZNRVBNU2zGyg_E:-L6BLZaiH01ykeD_oqXa:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/86b83bf532db83455d0010843f07040c86d89036/src/chatlog/textformatter.cpp#L185)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`!matchUri.valid`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4951)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4951#-:-L6DWXCeVFEUJH7qK29E:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
